TSRM/TSRM.c

```diff
@@ -170,6 +177,11 @@ TSRM_API void tsrm_shutdown(void)
 {
 	int i;
 
+	if (!in_main_thread) {
+		/* ensure singleton */
+		return;
+	}
+
 	if (tsrm_tls_table) {
```

Comment:
This seems ok, i've reverted for now as it's not related to interned strings directly. Will re-apply once the actual interned strings stuff is finished.